04-onion-routing.md

```diff
@@ -755,59 +763,153 @@ Note that these error messages are not reliable, as they are not placed on-chain
 due to the possibility of hop failure.
 
 Intermediate hops store the shared secret from the forward path and reuse it to
-obfuscate any corresponding return packet during each hop.
+authenticate and obfuscate any corresponding return packet during each hop.
 In addition, each node locally stores data regarding its own sending peer in the
 route, so it knows where to return-forward any eventual return packets.
-The node generating the error message (_erring node_) builds a return packet
+
+The origin node signals to each node on the path that it supports the error
+format that is detailed below by setting the `fat_error` tlv record in the
+forward packet to contain the desired structure. It must make sure that each
+node supports this format by observing the node feature `option_fat_error`.
+
+By choosing a structure for the fat errors, the origin node is making a
+trade-off between failure message size and privacy. For maximum privacy,
+`max_hops` should be set to the global maximum of 27. A lower value reveals a
+lower upper bound for the route length.
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`max_hops*(max_payload_size+1)*byte`:`payloads`]
+   * [`total_hmacs*32*byte`:`hmacs`]
+
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
+
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum number of hops in a route (`max_hops`) and the
+payload size (`max_payload_size`). An extra payload byte is added to the size
+signaled in the `fat_error` onion tlv for the payload type.
+
+Currently `max_payload_size` is fixed at 8. The field exists for future
+extensibility.
+
+The per-hop payload consists of the following fields:
+1. data:
+   * [`byte`:`payload_type`]
+   * [`u64`:`hold_time_ms`]
```

Comment:
Missed this comment during my review, TLV stream seems like the right approach to save us future headaches. 

> One issue with tlv is that there is no length indication. The payload will always be the max size.

This is an issue when the TLV payload is _less_ than the `max_payload_size`, right? Would a padding TLV to fill the space work here?

Reply:
Yes indeed, if it is less than `max_payload_size`. With padding you mean adding a dummy tlv field to fill all bytes? I think this can be problematic because not every length can be filled. It wouldn't be possible for example to fill a single remaining byte, because a tlv record is at least 2 bytes.

Reply:
> With padding you mean adding a dummy tlv field to fill all bytes?

Yes, seems to be the resolution reached [here](https://github.com/lightning/bolts/pull/607#issuecomment-496178924) (although I didn't follow that PR closely at the time). 

> I think this can be problematic because not every length can be filled. 

Indeed there are some edge cases here, but it could be enough to add the restriction that your "application level" TLVs must fit in `max_payload_size` -2?

Reply:
I think the resolution reached was to do no sentinel and no padding either?

>The critical thing is that the decoder knows where it should stop reading. In the case of the onion, we prepend the length directly to the tlv_stream, and for wire messages the packet length covers any existing fields and the tlv_stream.

To me the fat error payload problem mostly resembles the onion case, and it seems most consistent to also prepend the length to the fat error tlv stream? It should be a single byte, because of the varint encoding.

>Indeed there are some edge cases here, but it could be enough to add the restriction that your "application level" TLVs must fit in max_payload_size -2?

But this means that you always waste two bytes?